### PR TITLE
 SHIP-0037: Use build output timestamps

### PR DESCRIPTION
**Pull request by HeavyWombat**
Add SHIP to propose the introduction of output image timestamp processing.

### CODE REVIEW

**Review by thanhnho291203 — APPROVED** [review]
0037

File: ships/0037-build-output-timestamp.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+The same fields will be made available in a BuildRun's `spec.output` and `spec.buildSpec.output`.
+
+### Backend
+
+The timestamp processing will be performed as part of the [image-processing step that is introduced with Shipwright-managed push](0026-shipwright-managed-push.md#backend).
+
+It can only work with Shipwright managed push strategies.
+
+The Git source step needs to be extended with a flag so that it produces a result file for the commit timestamp, similar to the result files for author or commit SHA. This would only be used in case the `SourceTimestamp` is configured in the build.
```
Comment (SaschaSchwarze0) [inline ships/0037-build-output-timestamp.md:94]: Can you please clarify what other source types will do to produce a source timestamp? I see two options:

* We define how other source types create a timestamp, I think that should be possible.
* We define that the image-processing step fails when the source timestamp result is not set.

**Review by SaschaSchwarze0 — CHANGES_REQUESTED** [review]
Mostly fine. Just some little more clarification needed.

**Review by SaschaSchwarze0 — APPROVED** [review]
/approve

File: ships/0037-build-output-timestamp.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+#### Story 3
+
+As a Shipwright user, I want write a build that results in a build run that does not touch the image timestamp at all.
+
+## Proposal
+
+### API Changes
+
+We will extend the Build resource by an additional (Unix epoch) timestamp entry in `spec.output`:
```
Comment (qu1queee) [inline ships/0037-build-output-timestamp.md:71]: pls also include  on the  `BuildRunStatus` object, we need to know what will change there, .e.g.  how are we gonna surface the timestamp value to users via the BuildRun status?

File: ships/0037-build-output-timestamp.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+- `<timestamp>` must be a parsable Unix epoch timestamp that lies in the past and will be then used as-is for the creation timestamp of the image, i.e. custom neutral timestamp. An empty string is considered not parsable and therefore results in an error.
+- Null (unset) results in no timestamp usage at all, which means that the respective build strategy timestamp is used.
+
+The same fields will be made available in a BuildRun's `spec.output` and `spec.buildSpec.output`.
+
+### Backend
+
+The timestamp processing will be performed as part of the [image-processing step that is introduced with Shipwright-managed push](0026-shipwright-managed-push.md#backend).
+
+It can only work with Shipwright managed push strategies and it will rely on the source step providing the necessary timestamp piece of information. If not provided with a timestamp, it will fail.
```
Comment (qu1queee) [inline ships/0037-build-output-timestamp.md:92]: on failures, pls provide the proposed `Status`(should be False), the `Reason` and the `Description`. This should be part of the API proposed changes.

**Review by qu1queee — CHANGES_REQUESTED** [review]
Requesting some extensions on the API part.

File: ships/0037-build-output-timestamp.md (hunk 2 above)
Comment (HeavyWombat) [reply ships/0037-build-output-timestamp.md:71]: With regards to the spec, it is covered a few sentences later:
> The same fields will be made available in a BuildRun's `spec.output` and `spec.buildSpec.output`.

**Review by qu1queee — APPROVED** [review]
/lgtm

**Pull request merged**